tech/post/oslo_privsep.md

```diff
@@ -0,0 +1,176 @@
+```
+
+### Using oslo.privsep
+
+When using privsep in our service, we need to define the priveleged context and then use it as a decorator on our command execute code.
+
+#### Define privsep context
+```python 
+default = priv_context.PrivContext(
+    __name__,
+    cfg_section='privsep_osbrick',
+    pypath=__name__ + '.default',
+    capabilities=[c.CAP_SYS_ADMIN],
+).
+```
+By this code, we defined the configure section which will be used to initialize the deamon in the future and assign it within the required capabilities.  The configuration options are below:
+```
+[privsep_osbrick]
+user = novapriv
+group = novapriv
+# This will overwrite the default capabilitie those are specified in the initialize code.
+capabilities = CAP_SYS_ADMIN, CAP_NET_ADMIN 
+```
+When we execute the privelged command through privsep for the first time, privsep will start up the daemon process with the user, group and
+capabilities.
+
+### Use the privsep context
+It's easy to make one of your commands to run with desired capabilties by using the privsep's decorator, see the ``entrypoint`` below:
+```python
+@privileged.default.entrypoint
+def execute_root(*cmd, **kwargs):
+    """NB: Raises processutils.ProcessExecutionError/OSError on failure."""
+    return custom_execute(*cmd, shell=False, run_as_root=False, **kwargs)
+```
+And the whole magic comes from these two methods:
+```python
+    def entrypoint(self, func):
+        """This is intended to be used as a decorator."""
+
+        assert func.__module__.startswith(self.prefix), (
+            '%r entrypoints must be below "%s"' % (self, self.prefix))
+
+        # Right now, we only track a single context in
+        # _ENTRYPOINT_ATTR.  This could easily be expanded into a set,
+        # but that will increase the memory overhead.  Revisit if/when
+        # someone has a need to associate the same entrypoint with
+        # multiple contexts.
+        assert getattr(func, _ENTRYPOINT_ATTR, None) is None, (
+            '%r is already associated with another PrivContext' % func)
+
+        f = functools.partial(self._wrap, func)
+        setattr(f, _ENTRYPOINT_ATTR, self)
+        return f
+```
+and
+```python
+    def _wrap(self, func, *args, **kwargs):
+        if self.client_mode:
+            name = '%s.%s' % (func.__module__, func.__name__)
+            if self.channel is None:
+                self.start()
+            return self.channel.remote_call(name, args, kwargs)
+        else:
+            return func(*args, **kwargs)
+```
+With the transformation of the decorator any execute commands are been passed to the privsep daemon through ``linxu domain socket``. The daemon
```

Comment:
IMO, we could add a comma symbol here.  With the transformation of the decorator,